pyvista/core/utilities/geometric_objects.py

```diff
@@ -383,6 +395,526 @@ def Sphere(
+        Solid sphere mesh.
+
+    See Also
+    --------
+    pyvista.SolidSphere: Sphere creation using linear sampling.
+    pyvista.Sphere: Sphere that describes outer 2D surface.
+
+    Examples
+    --------
+    Linearly sampling spherical coordinates does not lead to
+    cells of all the same size at each radial position.
+    Cells near the poles have smaller sizes.
+
+    >>> import pyvista as pv
+    >>> import numpy as np
+    >>> solid_sphere = pv.SolidSphereGeneric(
+    ...     radius=np.linspace(0, 0.5, 2),
+    ...     theta=np.linspace(180, 360, 30),
+    ...     phi=np.linspace(0, 180, 30),
+    ... )
+    >>> solid_sphere = solid_sphere.compute_cell_sizes()
+    >>> solid_sphere.plot(
+    ...     scalars="Volume", show_edges=True, clim=[3e-5, 5e-4]
+    ... )
+
+    Sampling the polar angle in a nonlinear manner allows for consistent cell volumes.  See
+    `Sphere Point Picking <https://mathworld.wolfram.com/SpherePointPicking.html>`_.
+
+    >>> phi = np.rad2deg(np.arccos(np.linspace(1, -1, 30)))
+    >>> solid_sphere = pv.SolidSphereGeneric(
+    ...     radius=np.linspace(0, 0.5, 2),
+    ...     theta=np.linspace(180, 360, 30),
+    ...     phi=phi,
+    ... )
+    >>> solid_sphere = solid_sphere.compute_cell_sizes()
+    >>> solid_sphere.plot(
+    ...     scalars="Volume", show_edges=True, clim=[3e-5, 5e-4]
+    ... )
+
+    """
+    if radius is None:
+        radius = np.linspace(0, 0.5, 5)
+    radius = np.asanyarray(radius)
+
+    # Hereafter all degrees are in radians
+    # Also radius, phi, theta are now np.ndarrays
+    if theta is None:
+        theta = np.linspace(0, 2 * np.pi, 30)
+    else:
+        theta = np.asanyarray(theta) if radians else np.deg2rad(theta)
+
+    if phi is None:
+        phi = np.linspace(0, np.pi, 30)
+    else:
+        phi = np.asanyarray(phi) if radians else np.deg2rad(phi)
+
+    nr = len(radius)
+    ntheta = len(theta)
+    nphi = len(phi)
+
+    if nr < 2:
+        raise ValueError("radius resolution must be 2 or more")
+    if ntheta < 2:
+        raise ValueError("theta resolution must be 2 or more")
+    if nphi < 2:
+        raise ValueError("phi resolution must be 2 or more")
+
+    def _is_sorted(a):
+        return np.all(a[:-1] < a[1:])
+
+    if not _is_sorted(radius):
+        raise ValueError("radius is not monotonically increasing")
+    if not _is_sorted(theta):
+        raise ValueError("theta is not monotonically increasing")
+    if not _is_sorted(phi):
+        raise ValueError("phi is not monotonically increasing")
+
+    def _greater_than_equal_or_close(value1, value2, atol):
+        return value1 >= value2 or np.isclose(value1, value2, rtol=0.0, atol=atol)
+
+    def _less_than_equal_or_close(value1, value2, atol):
+        return value1 <= value2 or np.isclose(value1, value2, rtol=0.0, atol=atol)
+
+    if not _greater_than_equal_or_close(radius[0], 0.0, tol_radius):
+        raise ValueError("minimum radius cannot be negative")
```

Comment:
This seems to allow `radius[0] = 1e-9` with default tolerance. Did I see correctly that we never clip the values to 0? (Let's ignore the possibility that the radius starts like `[1e-10, 1e-9, 1e-7, 1e-6]` in which case the first two values would be accepted, and both clipped to 0, leading to singular values). I suspect this is what you mentioned in another comment about not "snapping" to 0 and friends, but having an invalid (negative) value here is more problematic.

Do things still work if we pass in a single too small (i.e. small negative) leading value? If not, we should probably test this (unless we already do).

Reply:
Sorry, I might have been imprecise before.  The way this currently works is that there are two types of tolerance.  Radius and angle.

For radius, if the first radius value is within the tolerance, we place an origin at [0, 0, 0], even if it is slightly negative.  I think this is fine.

For phi, if the first phi is within the tolerance to 0, then it puts the point there but doesn't snap the phi to 0.  Now that I think more about this, it results in a self-intersecting cell structure as this slightly off-phi rotates around the axis.  I am now convinced we need to do this snapping for phi.

For theta, we only detect overlap of start and endpoint.  In this case when overlapping occurs, we drop the endpoint and only use the start point.  So we do not need to snap, nor would I know how to do it.  There could be an issue when there are multiple points that overlap, but this seems like a user error at this point.

Reply:
> For phi, if the first phi is within the tolerance to 0, then it puts the point there but doesn't snap the phi to 0. Now that I think more about this, it results in a self-intersecting cell structure as this slightly off-phi rotates around the axis. I am now convinced we need to do this snapping for phi.

Specifically for this case, yes, I think we have to force non-negative polar angles. In my head this is less "snap" and more "clip": I don't mind small _positive_ polar angles staying like that.

Reply:
For radius: I'd expect anything that `isclose()` to 0 (whether it is positive or negative) to be "snapped" to 0. Anything that is negative and not close to 0 should raise an error. If radius > 0 and not close to 0 I'd expect a radius gap at the origin (otherwise no gap).

For phi (polar): I'd expect anything that `isclose()` to 0 (whether it is positive or negative) to be snapped to 0. Anything that is negative and not close to 0 should raise an error. If phi > 0 and not close to 0 I'd expect a phi gap at the north pole (otherwise no gap). I'd expect anything that `isclose()` to pi (whether it is less than or greater than pi) to be snapped to pi. Anything that is greater than pi and not close to pi should raise an error. If phi < pi and not close to pi I'd expect a phi gap at the south pole (otherwise no gap).

for theta (azimuthal): If theta[-1] is close to theta[0] (i.e. zero span in theta) I would expect an exception to be raised. If theta[-1] is close to theta[0] + 2pi (larger or smaller) then I'd expect theta[-1] to be snapped to theta[0] + 2pi (and for a compete gapless revolution to be applied. If theta[-1] is greater than theta[0] but not close I'd expect an exception to be raised. If theta[-1] is not close to theta[0] + 2pi then I would expect a theta gap (otherwise no gap).

I *think* these conditions, combined with the monotonically increasing check, should successfully validate/invalidate ALL user input. I don't think there's any edge cases missed in this case.

Reply:
I align with most of @jagerber48 suggestions here.  It makes sense that if we are going to have a tolerance, to snap it to the expected value for radius and phi.  This can be turned off by users by setting the tolerance to be 0 and taking control of precision themselves.

The only difference is that I think we can allow theta[-1] to be close in absolute terms to theta[0], e.g. 1-8 and 1.5e-8, since these don't overlap in the sense that they overlap near 2*pi away from each other.  There is no ambiguity that these are not supposed to be wrapped around and considered a full revolution.

I'm also unsure whether this line makes sense, but maybe it is modifying some other statement?

> If theta[-1] is greater than theta[0] but not close I'd expect an exception to be raised.

Reply:
@MatthewFlamm My understanding of the whole `isclose()` thing is to be a helper for users who might be passing in arrays generated by some numerical routine where really r[0] should be 0, or phi should be pi, but it's off by numerical precision. I think the same story applies to theta. The user might have theta generated by some numerical routine where theta[0] = 0 and theta[-1] SHOULD BE 2 pi, but it's off by a small numerical factor. In this case (is close) I think the code should just behave as if theta[-1] = theta[0] + 2 pi, i.e. there should be no theta gap in the generated mesh.

I don't understand your 1e-8 and 1.5e-8. If theta[0] = 1e-8 and theta[-1] = 1.5e-8 I think that should be an error because it's not monotonically increasing (with 1e-8 and 1.5e-8 being considered the same since they're within tolerance of each other)

I think that last comment you didn't understand was supposed to read
> if theta[-1] is greater than theta[0] **+ 2pi** but not close I'd expect an exception to be raised.

Reply:
In the first part, I think this is what we have currently.  It checks `np.isclose(theta[-1] - theta[0], 2 *np.pi, rtol=0, atol=tol_angle)`. If this is true, we treat them as the same point in 3D cartesian space and make the mesh totally seamless here. 

> I don't understand your 1e-8 and 1.5e-8. If theta[0] = 1e-8 and theta[-1] = 1.5e-8 I think that should be an error because it's not monotonically increasing (with 1e-8 and 1.5e-8 being considered the same since they're within tolerance of each other)

Since we only check congruence at 2pi I don't think we need to raise an error.  Someone may want to create a thin wedge.

>I think that last comment you didn't understand was supposed to read

> > if theta[-1] is greater than theta[0] + 2pi but not close I'd expect an exception to be raised.




The last example makes more sense to me now and it is the way it works currently.

Reply:
> @MatthewFlamm My understanding of the whole `isclose()` thing is to be a helper for users who might be passing in arrays generated by some numerical routine where really r[0] should be 0, or phi should be pi, but it's off by numerical precision. I think the same story applies to theta. The user might have theta generated by some numerical routine where theta[0] = 0 and theta[-1] SHOULD BE 2 pi, but it's off by a small numerical factor. In this case (is close) I think the code should just behave as if theta[-1] = theta[0] + 2 pi, i.e. there should be no theta gap in the generated mesh. That is, the point isn't to resolve some sort of ambiguity, but rather to provide a little bit of automatic input sanitization.

That's precisely my mental model. With the addition that most end users won't be familiar with the intricacies of floating-point math and be taken off guard by this. It would be hard for them to find the issue, and even then hard to fix it ("What do you mean my last value is not `pi`? I'm taking `n` steps of `pi/n`!").

Reply:
> > I don't understand your 1e-8 and 1.5e-8. If theta[0] = 1e-8 and theta[-1] = 1.5e-8 I think that should be an error because it's not monotonically increasing (with 1e-8 and 1.5e-8 being considered the same since they're within tolerance of each other)
> 
> Since we only check congruence at 2pi I don't think we need to raise an error. Someone may want to create a thin wedge.

I see, if a user passes `theta = [0, 1e-9]` you think a thin wedge should be created rather than an error. I guess I'm suggesting that whatever function checks for monotonic increasing would check that each point is more than `tolerance` away from its neighbor. something like

```
def is_monotonic_increasing(arr):
    if not np.all(arr[1:-1] > arr[0:-2]):
        return False
    if np.any(np.isclose(arr[1:1], arr[0:-2]):
        return False
    return True
```

In this case if the user passed in `theta = [0, 1e-9]` with `tolerance = 1e-8` they would get an error for having a non-monotonic-increasing array. Basically the blanket policy is that the `SolidSphereGeneric` function considers ANY numbers that are within `tolerance` to be the same number.

But yeah, this is not strictly necessary. It can be fine to have cells with sizes < tolerance within the field of the mesh (i.e. away from the origin/axes/prime meridian) as long as the edge cases at the edges of the mesh are handled well.

Yeah, I don't think I have a strong opinion on this. Whatever you think is good is fine to me. Perhaps my suggestion of considering `isclose` for purposes of detecting monotonic increasing is overly restrictive when it doesn't need to be, and `isclose` need only be considered at the edges.

Reply:
(I think one could also do `(arr[1:] > arr[:-1] + absolute_tolerance).all()`)